[#6] Enum과 ExceptionHandler를 활용한 공통 예외처리

msugo1 opened the pull request:
### `Exception Response` 클래스 작성
* 예외 메세지와 응답 코드를 담아서 클라리언트에 전달해주는 DTO 객체를
  만들기 위해 클래스 작성

* enum 클래스의 상수 객체를 주입받아서 해당 정보를 가지고 새로운 객체를
  생성하는 생성자 추가

### Http Status Code와 예외 관련한 메세지를 담을 `Enum`클래스 생성
* `ExceptionCode` (Interface)
: DTO에 Enum 상수객체를 주입할 때 클래스의 실제 타입에 상관없이 Interface 타입을 가지고 자유롭게 담을 수 있게 인터페이스를 구현

* `FileExceptionCode`, `LoginExceptionCode`, `UserExceptionCode` (new)
: 각 예외 상황과 관련된 Http Status Code와 설명을 담은 상수들을 작성 
(기존에 각 예외마다 일일이 만들었던 개별 예외클래스들 제거) 

### `공통 예외` 클래스 작성
- `FileException`, `LoginException`, `UserException` (new)
: 상황에 맞게 개별 예외를 담을 공통 예외클래스 작성

### `ExceptionHandler` 업데이트 (각 공통예외를 처리할 핸들러 메소드 추가)
- `logging`: 예외 메세지와 예외 객체를 함께 전달

- ResponseEntity 객체에 DTO 객체와 DTO 객체가 가지고 있는 상태코드를
  담아서 클라이언트에 리턴
  (@ResponseStatus로 특정 HttpStatus code를 지정하는 대신 예외 처리시
  가져온 상태코드를 담아서 각 예외에 맞는 유동적인 상태코드 리턴)

### 수정한 내용을 테스트 코드에 반영 및 테스트 구동

Commit:
[#6] Enum과 ExceptionHandler를 활용한 공통 예외처리

Exception Response 클래스 작성
* 예외 메세지와 응답 코드를 담아서 클라리언트에 전달해주는 DTO 객체를
  만들기 위해 클래스 작성

* enum 클래스의 상수 객체를 주입받아서 해당 정보를 가지고 새로운 객체를
  생성하는 생성자 추가

Http Status Code와 예외 관련한 메세지를 담을Enum 클래스 생성
* ExceptionCode (Interface)
- DTO에 Enum 상수객체를 주입할 때 클래스의 실제 타입에 상관없이 Interface 타입을 가지고 자유롭게 확장할 수 있도록 하는 인터페이스

* FileExceptionCode/LoginExceptionCode/UserExceptionCode (new)
- 각 예외 상황과 관련된 Http Status Code와 설명을 담은 상수들을 작성

공통 예외 클래스 작성
- FileException/LoginException/UserException (new)
: 각 상황에 맞는 예외를 담을 공통 예외클래스

ExceptionHandler 업데이트
* 각 공통예외를 처리할 핸들러 메소드 추가
- 로깅: 예외 관련 메세지와 예외 객체를 같이 전달
- ResponseEntity 객체에 DTO 객체와 DTO 객체가 가지고 있는 상태코드를
  담아서 클라이언트에 리턴
  (@ResponseStatus로 특정 HttpStatus code를 지정하는 대신 예외 처리시
  가져온 상태코드를 담았기에 유동적인 상태코드 리턴이 가능)

수정한 내용을 테스트 코드에 반영 및 테스트 구동

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/UserController.java:
```diff
@@ -44,18 +42,10 @@
-        try {
-            User user = userService.getUser(loginRequest);
-            loginService.login(user.getUserId());
+        User user = userService.getUser(loginRequest);
+        loginService.login(user.getUserId());
```
유저아이디로 유저를 불러오는것도 비지니스 로직이 아닐까요?

msugo1 replied:
저 부분 고민하다가 핵심 비즈니스 로직이 `loginService`고 유저 아이디를 통해 유저를 불러오는 부분은 보조적인 부분이라고 생각해서 저렇게 구현을 했거든요.

이후 `저번 이슈(#5)`에서 현재 유저 패스워드 변경시 로그아웃 시키는 부분을 서비스 계층으로 옮기고자 userService가 loginService를 주입받도록 수정했었더니 이제 `LoginService`에서 유저 아이디를 얻어오기 위해 유저를 UserService를 주입받으려면  순환참조가 일어나겠더라구요.

순환참조는 보통 설계가 잘못될 시 나타난다고 하는데 결국 설계 부분이 잘못된게 아닐까 하네요;

일단 제가 고민해본 결과 순환참조를 해결하기 위해서라도 저 부분이 필요해보이는데요. 아니면 저 부분을 서비스 계층으로 데려가려면 유저의 비밀번호가 업데이트되면 로그아웃되는 로직을 고집하지 않고 UserService가 주입받은 LoginService를 없앤 뒤 저 부분을 비즈니스 계층으로 데려갈 수 있을 것 같구요.  

구글에 검색을 해봐도 순환참조를 해결하려면 `redesign`을 하던지 아니면 `setter 주입` 혹은 `@lazy 로딩`을 사용하라고 하는데  후자는 아무리봐도 지금 상황만 해결하려는 미봉책인 것 같은데요. 혹시 헤쳐나가려면 방향성을 어떻게 잡는 것이 좋을까요?

msugo1 replied:
비밀번호 업데이트 후 로그아웃 기능을 제거하여 userService에서 loginService를 제거함으로써 해결하기로 했습니다! 이슈를 하나 더 파서 진행하도록 하겠습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/enums/LoginExceptionCode.java:
```diff
@@ -0,0 +1,16 @@
+
+@Getter
+@RequiredArgsConstructor
+public enum LoginExceptionCode implements ExceptionCode {
```
`ExceptionResponse` 클래스만 있으면 될 것 같은데 이쪽은 너무 오버엔지니어링 같습니다~
그리고 enum이 어떤 곳에 사용하는 것인지 좀 더 알아보시면 좋을 것 같네요

msugo1 replied:
너무 복잡하게 생각하다가 enum에 예외 추가하는 방식까지 가버렸었네요. 클래스 모두 제거하고 단순한 `RuntimeException`을 상속받은 예외 클래스들을 만드는 형식으로 바꾸었습니다.

여러 내용을 보며 정리한 Enum의 내용은 다음과 같은데요. 혹시 보충할 부분이나 틀린 부분 있으면 지적해주시면 감사하겠습니다.

### `enum`
- 서로 관련 있는 상수들을 모아 Symbolic한 명칭의 집합으로 정의한 것
- Enum 클래스형을 기반으로 한 클래스형 선언 (특정 타입을 가지고 있음)

### 활용
1. 데이터의 그룹화 및 관리를 위해
- 서로 관련되어 있지만 관련성을 표기하기 힘든 형태의 데이터를 한 곳에서 관리할 수 있다.
- Enum 사용시 보다 명확한 방법으로 이들의 관계를 가시적으로 표현 가능하다.

2. 상태와 행위를 한 곳에서 관리하기 위해
- enum도 결국엔 하나의 `class`이다.
- 특정 값과 메소드가 어떤 관계가 있는지를 한 곳에 표시가능하게 해준다.

f-lab-dev commented on src/main/java/me/soo/helloworld/util/ControllerExceptionHandler.java:
```diff
@@ -13,12 +13,25 @@
+        return new ResponseEntity<>(response, response.getStatus());
     }
+
+    @ExceptionHandler(LoginException.class)
```
`LoginException`, `UserException` 이렇게 보면 어떤 예외인지 명확하게 보이지 않습니다.
만약 `UserNotFoundException` 이렇다면 어떤 예외인지 명확해지고, 그렇기 때문에 Exception에 status code를 넣을 필요가 없어집니다

msugo1 replied:
Handler에서 예외를 공통된 이름으로 통합해서 처리하고 각 메소드에서 예외를 던질 때 명확하게 표시해주면 명확성도 잃지 않고 새로운 예외 클래스를 계속적으로 만들 필요가 없을 것이라 생각했는데 Handler에 추가해주는 예외들도 이름을 명확하게 해줘야겠네요. 공통예외를 상속해서 이거를 파라미터로 넘기고 상속받은 디테일한 예외들은 핸들러에  매핑해주는 방식으로 다시 전환하겠습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/exception/FileException.java:
```diff
@@ -0,0 +1,15 @@
+    private FileExceptionCode fileExceptionCode;
+
+    public FileException(FileExceptionCode userExceptionCode) {
+        this.fileExceptionCode = userExceptionCode;
```
status code는 http에 종속적인 스펙입니다. 이 예외와 값들은 어플리케이션 전체에서 쓰이는 것 같은데 그럼 어플리케이션이 http에 의존적이게 되지 않을까요?

msugo1 replied:
HttpStatus를 이용하는 대신 간단한 int 값을 속성으로 가지도록 변경하였습니다.

Commit:
[#6] #5에서의 업데이트를 반영하기 위한 Merge

- Merge branch 'feature/5' into feature/6

Commit:
[#6] 피드백 반영

* `오버엔지니어링` 문제 해결
- 인터페이스와 이를 상속 받은 enum클래스를 제거
- 이전처럼 `RuntimeException`을 상속받아 커스텀 예외 클래스를 만들어 활용하는
  형태로 회귀

* ExceptionHandler에 매핑된 예외이름의 명확성 문제 해결
- 기존: Handler에 추상적인 이름을 가진  공통 예외를 지정해서 어떤 예외인지 명확하게 파악할 수
  없었음
- 이름을 좀 더 세분화한 예외 클래스 생성
(UserException -> InvalidUserException, FileException ->
FileNotUploadedException, FileNotDeletedException)

* 각 예외 클래스와 ExceptionResponse가 Http에 종속적인 상태를 가졌던
  문제 해결
- HttpStatus를 활용하는 대신 단순 int 값을 활용하는 형태로 변경

Commit:
[#6] 피드백 반영

* UserController 계층에 존재하던 비즈니스 로직을 서비스 계층으로 이전
- 기존
  : UserController에서 UserService를 통해 User 정보를 얻어와서
  LoginService로 전달

- 변경
  : 비즈니스 로직인 User 정보를 얻어오는 부분을 loginService에서
  처리하도록 함

* 위의 변경에 따른 파라미터 타입 변경
LoginService

- login(String userId) -> login(UserLoginRequest -> userLoginRequest)
- getUser(UserLoginRequest userLoginRequest) -> getUser(String
  requestId, String requestPassword)

f-lab-dev commented on src/main/java/me/soo/helloworld/exception/InvalidUserException.java:
```diff
@@ -0,0 +1,12 @@
+package me.soo.helloworld.exception;
+
+public class InvalidUserException extends RuntimeException{
```
```suggestion
public class InvalidUserException extends RuntimeException {
```

f-lab-dev commented on src/main/java/me/soo/helloworld/model/ExceptionResponse.java:
```diff
@@ -1,18 +1,14 @@
 public class ExceptionResponse {
 
-    private HttpStatus status;
+    private final int status;
```
http status code는 헤더에 들어가는데 이 객체는 body에 들어가는 데이터이기때문에 중복 표현이 아닐까요?

Commit:
[#7] 피드백 반영

* Exception Handler
- 예외 response 시 중복되는 status code return 제거
- 대신 예외처리시 예외 처리에 대한 간단한 제목(title)과 세부사항(deatils)을 담아서 리턴

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/MyPageController.java:
```diff
@@ -38,7 +38,7 @@
     }
 
-    @PutMapping("/infos")
+    @PutMapping
```
이거는 중복작업 같습니다~ 부모 브랜치에서 작업한것을 중복으로 자식브랜치에서 작업할 필요는 없습니다. develop에 들어가면 다 똑같기 때문입니다.

Commit:
[#6] 테스트 수정

* MyPageControllerTest
- URL 변경에 따라 테스트 내의 URL도 변경 적용
